adyen/gateway.py

```diff
@@ -101,75 +103,82 @@ class Gateway:
         Constants.IDENTIFIER,
         Constants.SECRET_KEY,
         Constants.ACTION_URL,
+        Constants.HMAC_ALGORITHM,
     )
 
-    def __init__(self, settings=None):
+    def __init__(self, request, settings=None):
         """
         Initialize an Adyen gateway.
         """
         if settings is None:
             settings = {}
 
-        self.identifier = settings.get(Constants.IDENTIFIER)
-        self.secret_key = settings.get(Constants.SECRET_KEY)
-        self.action_url = settings.get(Constants.ACTION_URL)
-
-        if self.identifier is None or self.secret_key is None or self.action_url is None:
+        if not all([x in settings.keys() for x in self.MANDATORY_SETTINGS]):
             raise MissingParameterException(
                 "You need to specify the following parameters to initialize "
-                "the Adyen gateway: identifier, secret_key, action_url. "
-                "Please check your configuration."
+                "the Adyen gateway: %s. Please check your configuration." %
+                ", ".join(self.MANDATORY_SETTINGS)
             )
 
-    def _compute_hash(self, keys, params):
+        self.identifier = settings.get(Constants.IDENTIFIER)
+        self.secret_key = settings.get(Constants.SECRET_KEY)
+        self.action_url = settings.get(Constants.ACTION_URL)
+        self.hmac_algorithm = settings.get(Constants.HMAC_ALGORITHM)
+
+        if request:
+            # We must first find out whether this is a redirection or a notification.
+            self.request = request
+            if request.method == 'GET':
+                self.response_class = globals()['PaymentRedirection%s' % self.hmac_algorithm]
```

Comment:
And perhaps just to note, my intention here was to allow easy addition of further such modules simply by defining new classes, and then using "introspection" not having to then go and alter the code that consumes those classes.